## Welcome to mapDATAge

The mapDATAge package is designed to explore the presence of geographic and temporal patterns in ancient DNA data. It takes simple tabulated text files as input and contains different panels to draw a number of maps and plots that are common to most ancient DNA studies, including drawing: 
(1) the spatial and temporal distributions of a given set of samples (SMAP) or alleles (AMAP);
(2) temporal allelic trajectories (TRAJECTORY);
(3) maps of individual ancestry profiles on a map (ANCESTRY);
(4) PCA (or MDS) and map related maps of geographicspatio--temporal distributions (PCA);
(5) geographicspatio-temporal distributions of alleles at one or multiple loci (MULTIPLESNPS);
(6) (sub)haplogroup on adistribution maps (HAPLO), and;
(7) finally, options are provided to automatically generate figures applying a preselected range of temporal and spatial and temporal parameters, which can be useful to contrast data from different loci and/or species (ONECLICK). 
    mapDATAge is implemented and maintained in ShinyR by Dr Xuexue Liu at the Centre for Anthropobiology and Genomics of Toulouse (CAGT) (https://cagt.cnrs.fr/).
The underlying code is available at [https://github.com/xuefenfei712/mapDATAge].
Please cite the following publication [Xuexue Liu and Ludovic Orlando (2022). mapDATAge: a ShinyR package to chart ancient DNA data through space and time. Bioinformatics XX:XX-XX].
Please contact Dr Xuexue Liu (xuexue.liu@univ-tlse3.fr), if you have any questions, or suggestions for improvement.
